File: api-tsl/src/tr/gov/tubitak/uekae/esya/api/tsl/util/XmlUtil.java
package tr.gov.tubitak.uekae.esya.api.tsl.util;

import java.text.CharacterIterator;
import java.text.StringCharacterIterator;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.List;

import javax.xml.datatype.DatatypeConfigurationException;
import javax.xml.datatype.DatatypeFactory;
import javax.xml.datatype.XMLGregorianCalendar;

import org.w3c.dom.Attr;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.w3c.dom.Text;

import tr.gov.tubitak.uekae.esya.api.tsl.TSLException;
import tr.gov.tubitak.uekae.esya.api.tsl.model.core.NodeListImpl;

public class XmlUtil {
	public static Element getNextElement(Node el) {
		while ((el != null) && (el.getNodeType() != Node.ELEMENT_NODE)) {
			el = el.getNextSibling();
		}
		return (Element) el;

	}
	public static void addNSAttribute(Element aElement, String aPrefix,
			String aNamespace) {
		aElement.setAttributeNS("xmlns:" + aPrefix.trim(),
				Constants.NS_NAMESPACESPEC, aNamespace);
	}

	public static void removeChildren(Element aElement) {
		Node child;
		while ((child = aElement.getFirstChild()) != null) {
			aElement.removeChild(child);
		}
	}

	public static Element createDSctx(Document doc, String prefix, String namespace) {
		if ((prefix == null) || (prefix.trim().length() == 0)) {
			throw new IllegalArgumentException("You must supply a prefix");
		}

		Element ctx = doc.createElementNS(null, "namespaceContext");

		ctx.setAttributeNS(Constants.NS_NAMESPACESPEC, "xmlns:" +prefix.trim(), namespace);

		return ctx;
	}

	public static void addLineBreak(Element aElement) {
		aElement.appendChild(aElement.getOwnerDocument().createTextNode("\n"));
	}
    public static Element selectFirstElementNoNS(Node aSibling, String aNodeName)
    {
        while (aSibling != null) {
            if (aSibling.getLocalName()!=null && aSibling.getLocalName().equals(aNodeName)) {
                return (Element)aSibling;
            }
            aSibling = aSibling.getNextSibling();
        }
        return null;
    }

	// XML Dökümanı içindeki geçersiz XML karakterlerini kaldırır.
	// Geçerli olanlarla yer değiştirir.
	public static String escapeXMLData(String aText) {
		StringBuilder result = new StringBuilder();
		StringCharacterIterator iterator = new StringCharacterIterator(aText);
		char character = iterator.current();
		while (character != CharacterIterator.DONE) {
			if (character == '<') {
				result.append("&lt;");
			} else if (character == '>') {
				result.append("&gt;");
			} else if (character == '\"') {
				result.append("&quot;");
			} else if (character == '\'') {
				result.append("&#039;");
			} else if (character == '&') {
				result.append("&amp;");
			} else {
				// the char is not a special one
				// add it to the result as is
				result.append(character);
			}
			character = iterator.next();
		}
		return result.toString();
	}

	public static Element selectFirstElement(Node aSibling, String aNamespace,
			String aNodeName) {
		while (aSibling != null) {
			if (isNamespaceElement(aSibling, aNodeName, aNamespace)) {
				return (Element) aSibling;
			}
			aSibling = aSibling.getNextSibling();
		}
		return null;
	}

	public static boolean isNamespaceElement(Node el, String type, String ns) {
		return !((el == null) || el.getNamespaceURI() == null
				|| !el.getNamespaceURI().startsWith(ns) || !el.getLocalName()
				.equals(type));

	}

	public static Element[] selectNodes(Node aSibling, String aNamespace,
			String aNodeName) {
		int size = 20;
		Element[] a = new Element[size];
		int curr = 0;
		// List list=new ArrayList();
		while (aSibling != null) {
			if (isNamespaceElement(aSibling, aNodeName, aNamespace)) {
				a[curr++] = (Element) aSibling;
				if (size <= curr) {
					int cursize = size << 2;
					Element[] cp = new Element[cursize];
					System.arraycopy(a, 0, cp, 0, size);
					a = cp;
					size = cursize;
				}
			}
			aSibling = aSibling.getNextSibling();
		}
		Element[] af = new Element[curr];
		System.arraycopy(a, 0, af, 0, curr);
		return af;
	}

	public static Element findByIdAttr(Element aRoot, String aAttributeValue) {
		Element result = findByIdAttrRec(aRoot, aAttributeValue);
		if (result == null) {
			result = findLikeIdAttr(aRoot, aAttributeValue);
		}
		return result;
	}

	private static Element findByIdAttrRec(Element aRoot, String aAttributeValue) {
		String id = aRoot.getAttribute(Constants.ATTR_ID);
		if (id != null && id.equals(aAttributeValue)) {
			return aRoot;
		}
		NodeList elements = aRoot.getElementsByTagName("*");
		for (int i = 0; i < elements.getLength(); i++) {
			Element next = (Element) elements.item(i);
			Element found = findByIdAttr(next, aAttributeValue);
			if (found != null) {
				return found;
			}
		}
		return null;
	}

	private static final String[] IDs = { "id", "ID", "iD" };

	private static Element findLikeIdAttr(Element aRoot, String aAttributeValue) {
		for (int j = 0; j < IDs.length; j++) {
			String id = aRoot.getAttribute(IDs[j]);
			if (id != null && id.equals(aAttributeValue)) {
				return aRoot;
			}
		}
		NodeList elements = aRoot.getElementsByTagName("*");
		for (int i = 0; i < elements.getLength(); i++) {
			Element next = (Element) elements.item(i);
			Element found = findLikeIdAttr(next, aAttributeValue);
			if (found != null) {
				return found;
			}
		}
		return null;
	}

	public static String getText(Element element) {
		Node sibling = element.getFirstChild();
		StringBuilder sb = new StringBuilder();

		while (sibling != null) {
			if (sibling.getNodeType() == Node.TEXT_NODE) {
				Text t = (Text) sibling;
				sb.append(t.getData());
			}
			sibling = sibling.getNextSibling();
		}
		return sb.toString();
	}

	public static final String XML_PREAMBLE_STR = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n";
	public static final byte[] XML_PREAMBLE = XML_PREAMBLE_STR.getBytes();

	public static byte[] getBase64DecodedText(Element element)
			throws TSLException {
		String text = getText(element);
		byte[] bytes;
		try {
			bytes = Base64.decode(text.getBytes(), 0, text.getBytes().length);
			// bytes = Base64.decode(text);
		} catch (Exception x) {
			throw new TSLException(x, "errors.base64");
		}
		return bytes;
	}

	public static void setBase64EncodedText(Element aElement, byte[] aValue) {
		Node n = aElement.getFirstChild();
		while (n != null) {
			aElement.removeChild(n);
			n = n.getNextSibling();
		}

		if (aValue == null) {
			return;
		}

		String base64Encoded = Base64.encode(aValue);
		Text t = aElement.getOwnerDocument().createTextNode(base64Encoded);

		aElement.appendChild(t);
	}

	public static List<Element> selectChildElements(Element aElement) {
		List<Element> selected = new ArrayList<Element>();
		NodeList list = aElement.getChildNodes();

		for (int i = 0; i < list.getLength(); i++) {
			Node node = list.item(i);
			if (node.getNodeType() == Node.ELEMENT_NODE) {
				selected.add((Element) node);
			}
		}
		return selected;
	}

	public static NodeList getNodeSet(Node rootNode, boolean comments) {
		List<Node> list = new ArrayList<Node>();
		getSetRec(rootNode, list, comments);
		return new NodeListImpl(list);
	}

	private static void getSetRec(Node rootNode, List<Node> result,
			boolean comments) {
		switch (rootNode.getNodeType()) {
		case Node.ELEMENT_NODE:
			result.add(rootNode);
			Element el = (Element) rootNode;
			if (el.hasAttributes()) {
				NamedNodeMap nl = el.getAttributes();
				// XmlAttributeCollection nl = rootNode.Attributes;
				for (int i = 0; i < nl.getLength(); i++) {
					Node nlItem = nl.item(i);
					result.add(nlItem);

				}
			}
			// no return keep working
		case Node.DOCUMENT_NODE:
			for (Node r = rootNode.getFirstChild(); r != null; r = r
					.getNextSibling()) {
				if (r.getNodeType() == Node.TEXT_NODE) {
					result.add(r);
					while ((r != null) && (r.getNodeType() == Node.TEXT_NODE)) {
						r = r.getNextSibling();
					}
					if (r == null) {
						return;
					}
				}
				getSetRec(r, result, comments);
			}
			return;
		case Node.COMMENT_NODE:
			if (comments) {
				result.add(rootNode);
			}
			return;
		case Node.DOCUMENT_TYPE_NODE:
			return;
		default:
			result.add(rootNode);
			break;
		}
	}

	public static XMLGregorianCalendar createDate(Calendar aDate) {
		try {
			GregorianCalendar calendar = new GregorianCalendar();
			calendar.setTime(aDate.getTime());
			return DatatypeFactory.newInstance().newXMLGregorianCalendar(
					calendar);
		} catch (DatatypeConfigurationException e) {
			throw new Error(e);
		}
	}

	public static XMLGregorianCalendar getDate(Element aElement)
			throws TSLException {
		String text = getText(aElement);
		try {
			return DatatypeFactory.newInstance().newXMLGregorianCalendar(text);
		} catch (Throwable e) {
			throw new TSLException(e, "Incorrect date format: " + text);
		}
	}

	public static Document getOwnerDocument(NodeList aNodeList) {
		for (int i = 0; i < aNodeList.getLength(); i++) {
			Document doc;
			Node node = aNodeList.item(i);
			int nodeType = node.getNodeType();
			if (nodeType == Node.DOCUMENT_NODE) {
				doc = (Document) node;
			} else if (nodeType == Node.ATTRIBUTE_NODE) {
				doc = ((Attr) node).getOwnerElement().getOwnerDocument();
			} else
				doc = node.getOwnerDocument();

			if (doc != null)
				return doc;
		}
		return null;
	}
}

File: api-common/csharp/src/util/XmlUtil.cs
using System;
using System.Collections.Generic;
using System.Security;
using System.Text;
using System.Xml;

namespace tr.gov.tubitak.uekae.esya.api.common.util
{
    public static class XmlUtil
    {
        public static readonly String NS_NAMESPACESPEC     = "http://www.w3.org/2000/xmlns/";
        public static readonly String ATTR_ID                  = "Id";

        public static XmlElement getNextElement(XmlNode el)
        {
            while ((el != null) && (el.NodeType != XmlNodeType.Element))
            {
                el = el.NextSibling;
            }
            return (XmlElement)el;

        }

        public static void removeChildren(XmlElement aElement)
        {
            XmlNode child;
            while ((child = aElement.FirstChild) != null)
            {
                aElement.RemoveChild(child);
            }
        }

        public static void addLineBreak(XmlElement aElement)
        {
            if (aElement != null && aElement.OwnerDocument != null) aElement.AppendChild(aElement.OwnerDocument.CreateTextNode("\n"));
        }

        public static XmlElement createDSctx(XmlDocument doc, string prefix, string @namespace)
        {
            if ((prefix == null) || (prefix.Trim().Length == 0))
            {
                throw new ArgumentException("You must supply a prefix");
            }

            XmlElement ctx = doc.CreateElement("namespaceContext", null);

            //TODO Bu Constants.NS_NAMESPACESPEC olunca zaten namespace var hatası veriyor. .net
            //java tarafında nasıl oluyor anlamadım.
            //Ahmetle konuşmak lazım.
            //ctx.SetAttribute("xmlns:" + prefix.Trim(),Constants.NS_NAMESPACESPEC, @namespace);
            ctx.SetAttribute("xmlns:" + String.Intern(prefix).Trim(), @namespace);

            return ctx;
        }

        public static void addNSAttribute(XmlElement aElement, string aPrefix, string aNamespace)
        {
            aElement.SetAttribute("xmlns:" + aPrefix.Trim(), NS_NAMESPACESPEC, aNamespace);
        }

        // XML Dökümanı içindeki geçersiz XML karakterlerini kaldırır.
        //Geçerli olanlarla yer değiştirir.
        public static string escapeXMLData(string aText)
        {
            return SecurityElement.Escape(aText);
        }

        public static XmlElement findByIdAttr(XmlElement aRoot, string aAttributeValue)
        {
            XmlElement result = findByIdAttrRec(aRoot, aAttributeValue) ?? findLikeIdAttr(aRoot, aAttributeValue);
            return result;
        }


        private static XmlElement findByIdAttrRec(XmlElement aRoot, string aAttributeValue)
        {
            string id = aRoot != null ? aRoot.GetAttribute(ATTR_ID) : null;
            if (id != null && id == aAttributeValue)
            {
                return aRoot;
            }
            if (aRoot == null) return null;
            XmlNodeList elements = aRoot.GetElementsByTagName("*");
            for (int i = 0; i < elements.Count; i++)
            {
                XmlElement next = (XmlElement)elements.Item(i);
                XmlElement found = findByIdAttr(next, aAttributeValue);
                if (found != null)
                {
                    return found;
                }
            }
            return null;
        }

        private static readonly string[] IDs = { "id", "ID", "iD" };

        private static XmlElement findLikeIdAttr(XmlElement aRoot, string aAttributeValue)
        {
            foreach (string str in IDs)
            {
                if (aRoot == null) continue;
                string id = aRoot.GetAttribute(str);
                if (!string.IsNullOrEmpty(id) && id == aAttributeValue)
                {
                    return aRoot;
                }
            }
            if (aRoot == null) return null;
            XmlNodeList elements = aRoot.GetElementsByTagName("*");
            for (int i = 0; i < elements.Count; i++)
            {
                XmlElement next = (XmlElement)elements.Item(i);
                XmlElement found = findLikeIdAttr(next, aAttributeValue);
                if (found != null)
                {
                    return found;
                }
            }
            return null;
        }


        public static string getText(XmlNodeList aList)
        {
            StringBuilder buffer = new StringBuilder();
            if (aList == null) return buffer.ToString();

            for (int i = 0; i < aList.Count; i++)
            {
                XmlNode aNode = aList.Item(i);
                if (aNode == null || aNode.NodeType != XmlNodeType.Text) continue;
                XmlText t = (XmlText) aNode;
                buffer.Append(t.Data);
            }
            return buffer.ToString();
        }

        //*/

        public static string getText(XmlElement element)
        {
            if (element == null) return string.Empty;
            XmlNode sibling = element.FirstChild;
            StringBuilder sb = new StringBuilder();

            while (sibling != null)
            {
                if (sibling.NodeType == XmlNodeType.Text)
                {
                    XmlText t = (XmlText) sibling;
                    sb.Append(t.Data);
                }
                sibling = sibling.NextSibling;
            }
            return sb.ToString();
        }


        public static readonly string XML_PREAMBLE_STR= "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
        public static readonly byte[] XML_PREAMBLE = Encoding.UTF8.GetBytes(XML_PREAMBLE_STR);


        public static byte[] getBase64DecodedText(XmlElement element)
        {
            string text = getText(element);
            byte[] bytes;
            try
            {
                //bytes = Convert.FromBase64String(text); artik Base64 class'i var
                 bytes = Base64.Decode(text);
            }
            catch (Exception x)
            {
                throw new ESYARuntimeException("errors.base64", x);
            }
            return bytes;
        }

        public static void setBase64EncodedText(XmlElement aElement, byte[] aValue)
        {
            if (aElement != null)
            {
                XmlNode n = aElement.FirstChild;
                while (n != null)
                {
                    aElement.RemoveChild(n);
                    n = n.NextSibling;
                }
            }

            if (aValue == null)
            {
                return;
            }

            //string base64Encoded = Base64.Encode(aValue).ToString(); //Convert.ToBase64String(aValue);
            string base64Encoded = Convert.ToBase64String(aValue);
            if (aElement == null) return;
            if (aElement.OwnerDocument == null) return;
            XmlText t = aElement.OwnerDocument.CreateTextNode(base64Encoded);

            aElement.AppendChild(t);
        }


        public static XmlNodeList getNodeSet(XmlNode rootNode, bool comments)
        {
            IList<XmlNode> list = new List<XmlNode>();
            GetSetRec(rootNode, list, comments);
            return new NodeListImpl(list);
        }

        private static void GetSetRec(XmlNode rootNode, IList<XmlNode> result, bool comments)
        {
            if (rootNode == null) return;
            switch (rootNode.NodeType)
            {
                case XmlNodeType.Element:
                    result.Add(rootNode);
                    XmlElement el = (XmlElement)rootNode;
                    if (el.HasAttributes)
                    {
                        XmlAttributeCollection nl = el.Attributes;
                        for (int i = 0; i < nl.Count; i++)
                        {
                            result.Add(nl.Item(i));
                        }
                    }
                    GetSetRecForXmlNodeTypeDocument(rootNode, result, comments);
                    return;
                case XmlNodeType.Document:
                    GetSetRecForXmlNodeTypeDocument(rootNode, result, comments);
                    return;
                case XmlNodeType.Comment:
                    if (comments)
                    {
                        result.Add(rootNode);
                    }
                    return;
                case XmlNodeType.DocumentType:
                    return;
                default:
                    result.Add(rootNode);
                    break;
            }
        }

        private static void GetSetRecForXmlNodeTypeDocument(XmlNode rootNode, IList<XmlNode> result, bool comments)
        {
            for (XmlNode r = rootNode.FirstChild; r != null; r = r.NextSibling)
            {
                if (r.NodeType == XmlNodeType.Text)
                {
                    result.Add(r);
                    while ((r != null) && (r.NodeType == XmlNodeType.Text))
                    {
                        r = r.NextSibling;
                    }
                    if (r == null)
                    {
                        return;
                    }
                }
                GetSetRec(r, result, comments);
            }
        }

        public static IList<XmlElement> selectChildElements(XmlElement aElement)
        {
            IList<XmlElement> selected = new List<XmlElement>();
            if (aElement != null)
            {
                XmlNodeList list = aElement.ChildNodes;

                for (int i = 0; i < list.Count; i++)
                {
                    XmlNode node = list.Item(i);
                    if (node != null && node.NodeType == XmlNodeType.Element)
                    {
                        selected.Add((XmlElement)node);
                    }
                }
            }
            return selected;
        }

        public static XmlElement[] selectNodes(XmlNode aSibling, string aNamespace, string aNodeName)
        {
            int size = 20;
            XmlElement[] a = new XmlElement[size];
            int curr = 0;
            //List list=new ArrayList();
            while (aSibling != null)
            {
                if (isNamespaceElement(aSibling, aNodeName, aNamespace))
                {
                    a[curr++] = (XmlElement)aSibling;
                    if (size <= curr)
                    {
                        int cursize = size << 2;
                        XmlElement[] cp = new XmlElement[cursize];
                        Array.Copy(a, 0, cp, 0, size);
                        a = cp;
                        size = cursize;
                    }
                }
                aSibling = aSibling.NextSibling;
            }
            XmlElement[] af = new XmlElement[curr];
            Array.Copy(a, 0, af, 0, curr);
            return af;
        }

        public static XmlElement selectFirstElement(XmlNode aSibling, string aNamespace, string aNodeName)
        {
            while (aSibling != null)
            {
                if (isNamespaceElement(aSibling, aNodeName, aNamespace))
                {
                    return (XmlElement)aSibling;
                }
                aSibling = aSibling.NextSibling;
            }
            return null;
        }

        public static XmlElement selectFirstElementNoNS(XmlNode aSibling, string aNodeName)
        {
            while (aSibling != null)
            {
                if (aSibling.LocalName == aNodeName)
                {
                    return (XmlElement)aSibling;
                }
                aSibling = aSibling.NextSibling;
            }
            return null;
        }

        public static bool isNamespaceElement(XmlNode el, string type, string ns)
        {
            return !((el == null) || string.IsNullOrEmpty(el.NamespaceURI) || !el.NamespaceURI.StartsWith(ns) || (!string.IsNullOrEmpty(el.LocalName) && el.LocalName != type));
        }

        public static XmlDocument getOwnerDocument(HashSet<XmlNode> xpathNodeSet)
        {
            if (xpathNodeSet == null) return null;
            using (IEnumerator<XmlNode> enumerator = xpathNodeSet.GetEnumerator())
            {
                while (enumerator.MoveNext())
                {
                    XmlNode node = enumerator.Current;
                    if (node != null && node.NodeType == XmlNodeType.Document)
                    {
                        return (XmlDocument)node;
                    }
                    XmlDocument doc = null;
                    if (node != null && node.NodeType == XmlNodeType.Attribute)
                    {
                        var ownerDocument = ((XmlAttribute)node).OwnerDocument;
                        if (ownerDocument != null)
                            doc = ownerDocument.OwnerDocument;
                    }
                    else
                    {
                        if (node != null) doc = node.OwnerDocument;
                    }
                    if (doc != null)
                        return doc;
                }
            }
            return null;
        }

        public static HashSet<XmlNode> convertNodelistToSet(XmlNodeList xpathNodeSet)
        {

            if (xpathNodeSet == null)
            {
                return new HashSet<XmlNode>();
            }

            int length = xpathNodeSet.Count;
            HashSet<XmlNode> set = new HashSet<XmlNode>();
            for (int i = 0; i < length; i++)
            {
                set.Add(xpathNodeSet.Item(i));
            }

            return set;
        }

        public static string datetime2xmlgregstr(DateTime dt)
        {
            return XmlConvert.ToString(dt, XmlDateTimeSerializationMode.Local);
        }

        public static DateTime getDate(XmlElement aElement)
        {
            string text = getText(aElement);
            try
            {
                //TODO Bu kısmın test edilmesi gerekiyor.
                //ISO 8601 format
                return XmlConvert.ToDateTime(text, XmlDateTimeSerializationMode.Local);
                //as return DateTime.Parse(text,  null, DateTimeStyles.RoundtripKind);
            }
            catch (Exception e)
            {
                throw new ESYARuntimeException("Incorrect date format: " + text, e);
            }
        }
        public static string FindXPath(XmlNode node)
        {
            StringBuilder builder = new StringBuilder();
            while (node != null)
            {
                switch (node.NodeType)
                {
                    case XmlNodeType.Attribute:
                        builder.Insert(0, "/@" + node.Name);
                        node = ((XmlAttribute) node).OwnerElement;
                        break;
                    case XmlNodeType.Element:
                        int index = FindElementIndex((XmlElement) node);
                        builder.Insert(0, "/" + node.Name + "[" + index + "]");
                        node = node.ParentNode;
                        break;
                    case XmlNodeType.Document:
                        return builder.ToString();
                    default:
                        throw new ArgumentException("Only elements and attributes are supported");
                }
            }
            throw new ArgumentException("Node was not in a document");
        }

        static int FindElementIndex(XmlElement element)
        {
            if (element == null) throw new ArgumentException("Couldn't find element within parent");
            XmlNode parentNode = element.ParentNode;
            if (parentNode is XmlDocument)
            {
                return 1;
            }
            XmlElement parent = (XmlElement) parentNode;
            int index = 1;
            if (parent == null) throw new ArgumentException("Couldn't find element within parent");
            foreach (XmlNode candidate in parent.ChildNodes)
            {
                if (!(candidate is XmlElement) || candidate.Name != element.Name) continue;
                if (candidate == element)
                {
                    return index;
                }
                index++;
            }
            throw new ArgumentException("Couldn't find element within parent");
        }


        /* @orcun.ertugrul; XmlDsigEnvelopedSignatureTransform.GetOutputFromNode fonksiyonunda Path aracılığı ile node seçmek gerekiyordu. 
         * Framework içindeki XmlNode.SelectSingleNode fonksiyonu XmlNamespaceManager nesnesi istiyor. Bu nesnenin yaratılması için yeterli bilgi 
         * erişilmek istenen fonksiyonun içinde olmadığı için aşağıdaki SelectNodeFromPath fonksiyonu yazıldı.
         * Eğer kullanabiliyorsanız Framework içindeki XmlNode.SelectSingleNode fonksiyonunu kullanınız.
         */

        public static XmlNode SelectNodeFromPath(XmlNode retNode, string path)
        {
            String[] nodeNames = path.Split('/');
            XmlNode travelerNode = retNode;

            foreach (string nodePath in nodeNames)
            {
                if (string.IsNullOrEmpty(nodePath))
                    continue;

                string nodeName = nodePath.Substring(0, nodePath.LastIndexOf("[", StringComparison.Ordinal));
                string indexStr = nodePath.Substring(nodePath.LastIndexOf("[", StringComparison.Ordinal) + 1,
                    nodePath.LastIndexOf("]", StringComparison.Ordinal) - nodePath.LastIndexOf("[", StringComparison.Ordinal) - 1);

                int nodeIndex;
                int.TryParse(indexStr, out nodeIndex);

                int index = 0;
                if (travelerNode == null) continue;
                XmlNodeList childNodes = travelerNode.ChildNodes;

                foreach (XmlNode childNode in childNodes)
                {
                    if (childNode.Name == nodeName)
                        index ++;
                    if (nodeIndex != index) continue;
                    travelerNode = childNode;
                    break;
                }
            }

            return travelerNode;

        }
    }
}
